x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/DataTierMigrationRoutedStep.java

```diff
@@ -0,0 +1,118 @@
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Set;
+
+import static org.elasticsearch.xpack.cluster.routing.allocation.DataTierAllocationDecider.INDEX_ROUTING_INCLUDE_SETTING;
+import static org.elasticsearch.xpack.core.ilm.AllocationRoutedStep.getPendingAllocations;
+
+/**
+ * Checks whether all shards have been correctly routed in response to updating the allocation rules for an index in order
+ * to migrate the index to a new tier.
+ */
+public class DataTierMigrationRoutedStep extends ClusterStateWaitStep {
+    public static final String NAME = "check-migration";
+
+    private static final Logger logger = LogManager.getLogger(DataTierMigrationRoutedStep.class);
+
+    private static final Set<Setting<?>> ALL_CLUSTER_SETTINGS;
+
+    static {
+        Set<Setting<?>> allSettings = new HashSet<>(ClusterSettings.BUILT_IN_CLUSTER_SETTINGS);
+        allSettings.add(DataTierAllocationDecider.CLUSTER_ROUTING_REQUIRE_SETTING);
+        allSettings.add(DataTierAllocationDecider.CLUSTER_ROUTING_INCLUDE_SETTING);
+        allSettings.add(DataTierAllocationDecider.CLUSTER_ROUTING_EXCLUDE_SETTING);
+        ALL_CLUSTER_SETTINGS = allSettings;
+    }
+
+    private static final AllocationDeciders ALLOCATION_DECIDERS = new AllocationDeciders(
+        List.of(
+            new DataTierAllocationDecider(new ClusterSettings(Settings.EMPTY, ALL_CLUSTER_SETTINGS))
+        )
+    );
+
+    DataTierMigrationRoutedStep(StepKey key, StepKey nextStepKey) {
+        super(key, nextStepKey);
+    }
+
+    @Override
+    public boolean isRetryable() {
+        return true;
+    }
+
+    @Override
+    public Result isConditionMet(Index index, ClusterState clusterState) {
+        IndexMetadata idxMeta = clusterState.metadata().index(index);
+        if (idxMeta == null) {
+            // Index must have been since deleted, ignore it
+            logger.debug("[{}] lifecycle action for index [{}] executed but index no longer exists", getKey().getAction(), index.getName());
+            return new Result(false, null);
+        }
+        if (ActiveShardCount.ALL.enoughShardsActive(clusterState, index.getName()) == false) {
+            logger.debug("[{}] lifecycle action for index [{}] cannot make progress because not all shards are active",
+                    getKey().getAction(), index.getName());
+            return new Result(false, AllocationInfo.waitingForActiveShardsAllocationInfo(idxMeta.getNumberOfReplicas()));
+        }
+
+        int allocationPendingAllShards = getPendingAllocations(index, ALLOCATION_DECIDERS, clusterState);
+
+        if (allocationPendingAllShards > 0) {
+            String tier = INDEX_ROUTING_INCLUDE_SETTING.get(idxMeta.getSettings());
+            boolean targetTierNodeFound = false;
+            for (DiscoveryNode node : clusterState.nodes()) {
+                for (DiscoveryNodeRole role : node.getRoles()) {
+                    if (role.roleName().equals(tier)) {
```

Comment:
I don't think this will work, because it's checking for the exact role, so for example, if the allocation were set to `data_warm`, and we had `data` nodes but not nodes with the exact `data_warm` role, this would say that they were not found (even though they are).

We should check for either an exact match or the presence of any nodes with the `data` role

Reply:
Great point